### PR TITLE
Add diagnostic logs for OIDC flows

### PR DESCRIPTION
**Pull request by chamathns**
### Proposed changes in this pull request
Includes https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1696

Relates to wso2/product-is#12436

Depends on 
https://github.com/wso2/carbon-kernel/pull/3066
https://github.com/wso2/carbon-identity-framework/pull/3711

### When should this PR be merged
After https://github.com/wso2/carbon-kernel/pull/3066
https://github.com/wso2/carbon-identity-framework/pull/3711

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 1]
```diff
@@ -539,17 +557,23 @@ during post consent handling to get the original ConsentClaimsData object (Assum
...
             } else {
+                OAuth2Util.log("oauth-inbound-service", params, "SUCCESS", null , "hand-over-to-consent-service", null);
                 getSSOConsentService().processConsent(approvedClaimIds, serviceProvider,
                         loggedInUser, value, false);
             }
 
         } catch (OAuthSystemException | SSOConsentServiceException e) {
             String msg = "Error while processing consent of user: " + loggedInUser.toFullQualifiedUsername() + " for " +
                     "client_id: " + clientId + " of tenantDomain: " + spTenantDomain;
+            OAuth2Util.log("oauth-inbound-service", null, "FAILED", "System error occurred." , "process-consent", null);
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:576]: Don't we need same for ClaimMetadataException and RequestObjectException

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java [hunk 2]
```diff
@@ -4312,4 +4347,11 @@ public static String resolveTenantDomain(HttpServletRequest request) {
         }
         return IdentityTenantUtil.getTenantDomainFromContext();
     }
+
+    public static void log(String componentId, Map<String, Object> input, String resultStatus, String resultMessage,
+                            String actionId, Map<String, Object> configurations) {
+
+        LoggerUtils.triggerDiagnosticLogEvent(componentId, input, resultStatus, resultMessage, actionId,
```
Comment (senthalan) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java:4354]: Will be have different componentId using this util other than `"oauth-inbound-service"` ? in that case, we can remove it from parameter and define it here

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java (hunk 1 above)
Comment (chamathns) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java]: +1. added with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1737/commits/eef37f7d62f63b732a61665b40af7c5118f00268

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java (hunk 2 above)
Comment (chamathns) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java]: No, all the logs are having `"oauth-inbound-service"` as the componentId. +1. Addressed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1737/commits/ba650f85362ebaa25fbeec0b1716bccb3e5aaced

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 3]
```diff
@@ -539,21 +562,49 @@ during post consent handling to get the original ConsentClaimsData object (Assum
...
             }
 
+            Map<String, Object> params;
             if (hasPromptContainsConsent(oauth2Params)) {
+                if (OAuth2Util.isDiagnosticLogsEnabled()) {
+                    params = new HashMap<>();
+                    params.put("clientId", clientId);
+                    params.put("prompt", oauth2Params.getPrompt());
+                    OAuth2Util.log(params, "SUCCESS", null, "hand-over-to-consent-service",
+                            null);
```
Comment (pulasthi7) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:572]: can merge to 1 line, please check others as well

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 4]
```diff
@@ -2112,16 +2388,40 @@ private String handlePreConsent(OAuth2Parameters oauth2Params, AuthenticatedUser
...
+                    OAuth2Util.log(params, "SUCCESS", "Consent claims generation successful. " +
+                                    "Consent claims query parameter: " + additionalQueryParam, "generate-consent-claims",
+                            configs);
+                }
                 additionalQueryParam = buildQueryParamString(requestClaimsQueryParam, mandatoryClaimsQueryParam);
             }
         } catch (UnsupportedEncodingException | SSOConsentServiceException e) {
             String msg = "Error while handling user consent for claim for user: " + user.toFullQualifiedUsername() +
                     " for client_id: " + clientId + " of tenantDomain: " + spTenantDomain;
+            if (OAuth2Util.isDiagnosticLogsEnabled()){
```
Comment (pulasthi7) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:2409]: fix formatting

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/revoke/OAuthRevocationEndpoint.java [hunk 5]
```diff
@@ -89,12 +95,31 @@ public Response revokeAccessToken(@Context HttpServletRequest request, Multivalu
...
+            if (MapUtils.isNotEmpty(paramMap)) {
+                paramMap.forEach((key, value) -> {
+                    if (TOKEN_PARAM.equals(key) && CollectionUtils.isNotEmpty(value)) {
+                        params.put("token", value.get(0).replaceAll(".", "*"));
+                    } else {
+                        params.put(key, value);
+                    }
+                });
+            }
+            if (OAuth2Util.isDiagnosticLogsEnabled()) {
```
Comment (pulasthi7) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/revoke/OAuthRevocationEndpoint.java:108]: this should be checked prior to creating params map

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/OAuth2Service.java [hunk 6]
```diff
@@ -105,6 +105,9 @@ public OAuth2AuthorizeRespDTO authorize(OAuth2AuthorizeReqDTO oAuth2AuthorizeReq
                     AuthorizationHandlerManager.getInstance();
             return authzHandlerManager.handleAuthorization(oAuth2AuthorizeReqDTO);
         } catch (Exception e) {
+            if (OAuth2Util.isDiagnosticLogsEnabled()) {
+                OAuth2Util.log(null, "FAILED", "System error occurred.", "authorize-client", null);
```
Comment (pulasthi7) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/OAuth2Service.java:109]: No need to check if enabled, if you only use constant strings and already available values

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/revoke/OAuthRevocationEndpoint.java (hunk 5 above)
Comment (chamathns) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/revoke/OAuthRevocationEndpoint.java:108]: the params map is being used at L120 as well. Therefore, can't include the param map in the if block. We can always initiate an empty hashMap and add the values after a check. Is that a better approach?

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 7]
```diff
@@ -428,6 +429,20 @@ private static String getFormPostRedirectPage() {
     private Response handleResponseFromConsent(OAuthMessage oAuthMessage) throws OAuthSystemException,
             URISyntaxException, ConsentHandlingFailedException {
 
+        if (OAuth2Util.isDiagnosticLogsEnabled()) {
```
[the remark this answers is not in the document]
Comment (chamathns) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java]: +1. Addressed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1737/commits/1a6cec2c39bfaefd7007ff99c535b33ff07a8c36

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 8]
```diff
@@ -428,6 +429,20 @@ private static String getFormPostRedirectPage() {
...
+            if (oAuthMessage.getRequest() != null && MapUtils.isNotEmpty(oAuthMessage.getRequest().getParameterMap())) {
+                oAuthMessage.getRequest().getParameterMap().forEach((key, value) -> {
+                    if (ArrayUtils.isNotEmpty(value)) {
+                        params.put(key, Arrays.asList(value));
+                    }
+                });
+            }
+            OAuth2Util.log(params, "SUCCESS", "Successfully received consent response",
+                    "receive" +
+                            "-consent-response", null);
```
[the remark this answers is not in the document]
Comment (chamathns) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java]: Addressed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1737/commits/f7b766bb97ea3607cbc320c6d7b284f05d5668ba

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java (hunk 3 above)
Comment (chamathns) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java]: Addressed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1737/commits/f7b766bb97ea3607cbc320c6d7b284f05d5668ba

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (chamathns) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java]: Addressed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1737/commits/f7b766bb97ea3607cbc320c6d7b284f05d5668ba

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java (hunk 3 above)
Comment (chamathns) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java]: Addressed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1737/commits/f7b766bb97ea3607cbc320c6d7b284f05d5668ba

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java (hunk 4 above)
Comment (chamathns) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java]: Addressed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1737/commits/f7b766bb97ea3607cbc320c6d7b284f05d5668ba

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/OAuth2Service.java (hunk 6 above)
Comment (chamathns) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/OAuth2Service.java]: Addressed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1737/commits/f7b766bb97ea3607cbc320c6d7b284f05d5668ba

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/OIDCRequestObjectUtil.java [hunk 9]
```diff
@@ -105,29 +115,59 @@ private static void validateRequestObjectSignature(OAuth2Parameters oAuth2Parame
...
+                    validateSignature(oAuth2Parameters, requestObject, requestObjectValidator);
+                } else {
+                    // If request object is not signed we need to throw an exception.
+                    if (OAuth2LogsUtil.isDiagnosticLogsEnabled()) {
+                        Map<String, Object> params = new HashMap<>();
+                        params.put("clientId", clientId);
+
+                        Map<String, Object> configs = new HashMap<>();
+                        configs.put("requestObjectSignatureValidationEnabled", "true");
+                        OAuth2LogsUtil.log(params, "FAILED",
```
Comment (pulasthi7) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/OIDCRequestObjectUtil.java:139]: define "FAILED" as constant and re-use

Comment (chamathns) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/OIDCRequestObjectUtil.java]: addressed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1737/commits/26ac12223c14d5da6faf1a81dffa97cc14eee7bc

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dto/OAuth2AuthorizeReqDTO.java [hunk 10]
```diff
@@ -133,7 +133,12 @@ public void setConsumerKey(String consumerKey) {
     }
 
     public String[] getScopes() {
-        return scopes.clone();
+
+        if (scopes != null) {
+            return scopes.clone();
+        } else {
+            return null;
```
[the remark this answers is not in the document]
Comment (chamathns) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dto/OAuth2AuthorizeReqDTO.java]: addressed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1737/commits/26ac12223c14d5da6faf1a81dffa97cc14eee7bc

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/AccessTokenIssuer.java [hunk 11]
```diff
@@ -400,6 +473,13 @@ public OAuth2AccessTokenRespDTO issue(OAuth2AccessTokenReqDTO tokenReqDTO)
             log.debug("Access token issued to client Id: " + tokenReqDTO.getClientId() + " username: " +
                     tokReqMsgCtx.getAuthorizedUser() + " and scopes: " + tokenRespDTO.getAuthorizedScopes());
         }
+        Map<String, Object> params = new HashMap<>();
```
Comment (pulasthi7) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/AccessTokenIssuer.java:476]: move the map creation to the if block, do the same at other places as well

Comment (chamathns) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/AccessTokenIssuer.java]: +1. addressed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1737/commits/b11a186a08caa35286029e46bba6f115aa4ce711

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/OAuth2Service.java [hunk 12]
```diff
@@ -442,6 +534,21 @@ public OAuthRevocationResponseDTO revokeTokenByOAuthClient(OAuthRevocationReques
                         if ((OAuth2Util.getAppInformationByClientId(accessTokenDO.getConsumerKey()).
                                 isTokenBindingValidationEnabled()) && (!isValidTokenBinding(accessTokenDO.
                                 getTokenBinding(), revokeRequestDTO.getRequest()))) {
+                            Map<String, Object> params = new HashMap<>();
+                            params.put("clientId", accessTokenDO.getConsumerKey());
+                            if (accessTokenDO.getTokenBinding() != null) {
+                                params.put("tokenBindingType", accessTokenDO.getTokenBinding().getBindingType());
+                                params.put("tokenBindingValue", accessTokenDO.getTokenBinding().getBindingValue());
+                            }
```
[the remark this answers is not in the document]
Comment (chamathns) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/OAuth2Service.java]: addressed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1737/commits/b11a186a08caa35286029e46bba6f115aa4ce711

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/OAuth2Service.java [hunk 13]
```diff
@@ -395,6 +455,31 @@ public OAuthRevocationResponseDTO revokeTokenByOAuthClient(OAuthRevocationReques
...
+                                }
+                            }  else if (OAuthConstants.TokenStates.TOKEN_STATE_INACTIVE
+                                    .equals(refreshTokenDO.getRefreshTokenState())) {
+                                if (LoggerUtils.isDiagnosticLogsEnabled()) {
+                                    LoggerUtils.triggerDiagnosticLogEvent(
+                                            OAuthConstants.LogConstants.OAUTH_INBOUND_SERVICE, params,
+                                            OAuthConstants.LogConstants.SUCCESS, "Provided token is in inactive state.",
+                                            "revoke-token", null);
+                                }
+                            }
```
[the remark this answers is not in the document]
Comment (chamathns) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/OAuth2Service.java:479]: +1. addressed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1737/commits/b11a186a08caa35286029e46bba6f115aa4ce711

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 14]
```diff
@@ -1731,6 +1975,7 @@ private void validateRequestObjectParams(OAuthAuthzRequest oauthRequest) throws
         // With in the same request it can not be used both request parameter and request_uri parameter.
         if (StringUtils.isNotEmpty(oauthRequest.getParam(REQUEST)) && StringUtils.isNotEmpty(oauthRequest.getParam
                 (REQUEST_URI))) {
+            //diag log
```
[the remark this answers is not in the document]
Comment (chamathns) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java]: added a diag log here with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1737/commits/b11a186a08caa35286029e46bba6f115aa4ce711

**Review by pulasthi7 — APPROVED** [review]

**Pull request merged**